File: examples/react-greeting-app/test/greeting.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { buildGreeting, GreetingError, SUPPORTED_LOCALES } from '../src/domain/greeting.js';

test('saluda en español por defecto y recorta espacios', () => {
  assert.equal(buildGreeting('  Ada  '), 'Hola, Ada.');
});

test('saluda en inglés cuando locale=en', () => {
  assert.equal(buildGreeting('Grace', 'en'), 'Hello, Grace.');
});

test('acepta nombres con acentos, apóstrofos y guiones', () => {
  assert.equal(buildGreeting("José-María O'Neill"), "Hola, José-María O'Neill.");
});

test('rechaza nombre vacío', () => {
  assert.throws(() => buildGreeting(''), GreetingError);
  assert.throws(() => buildGreeting('   '), GreetingError);
});

test('rechaza caracteres fuera de la lista permitida (inyección)', () => {
  for (const bad of ['<script>', 'Ada; DROP TABLE', 'a\nb', '{{x}}']) {
    assert.throws(() => buildGreeting(bad), GreetingError);
  }
});

test('rechaza nombres de más de 80 caracteres', () => {
  assert.throws(() => buildGreeting('a'.repeat(81)), GreetingError);
});

test('rechaza idiomas no soportados y publica los soportados', () => {
  assert.deepEqual(SUPPORTED_LOCALES, ['es', 'en']);
  assert.throws(() => buildGreeting('Ada', 'fr'), GreetingError);
});

File: examples/flutter-greeting-app/test/greeting_test.dart
import 'package:flutter_greeting_app/domain/greeting.dart';
import 'package:test/test.dart';

void main() {
  group('buildGreeting', () {
    test('saluda en español por defecto', () {
      expect(buildGreeting('Ana'), 'Hola, Ana.');
    });

    test('saluda en inglés cuando se pide', () {
      expect(buildGreeting('Ada', locale: 'en'), 'Hello, Ada.');
    });

    test('recorta espacios al inicio y al final', () {
      expect(buildGreeting('  Ana  '), 'Hola, Ana.');
    });

    test('acepta apóstrofes y guiones', () {
      expect(buildGreeting("O'Brien"), "Hola, O'Brien.");
      expect(buildGreeting('Jean-Paul'), 'Hola, Jean-Paul.');
    });

    test('acepta nombres con tildes y eñes', () {
      expect(buildGreeting('María José'), 'Hola, María José.');
    });

    test('rechaza nombre nulo', () {
      expect(() => buildGreeting(null), throwsA(isA<GreetingError>()));
    });

    test('rechaza nombre vacío', () {
      expect(() => buildGreeting(''), throwsA(isA<GreetingError>()));
    });

    test('rechaza nombre solo con espacios', () {
      expect(() => buildGreeting('   '), throwsA(isA<GreetingError>()));
    });

    test('rechaza nombre con dígitos', () {
      expect(() => buildGreeting('Ana3'), throwsA(isA<GreetingError>()));
    });

    test('rechaza nombre con símbolos no permitidos', () {
      expect(() => buildGreeting('Ana@'), throwsA(isA<GreetingError>()));
    });

    test('rechaza nombre que excede el máximo de caracteres', () {
      final longName = 'A'.padRight(81, 'A');
      expect(() => buildGreeting(longName), throwsA(isA<GreetingError>()));
    });

    test('acepta nombre justo en el límite de caracteres', () {
      final maxName = 'A'.padRight(80, 'A');
      expect(buildGreeting(maxName), 'Hola, $maxName.');
    });

    test('rechaza locale no soportado', () {
      expect(
        () => buildGreeting('Ana', locale: 'fr'),
        throwsA(isA<GreetingError>()),
      );
    });

    test('el mensaje de error es legible, no una excepción genérica', () {
      expect(
        () => buildGreeting(''),
        throwsA(
          isA<GreetingError>().having(
            (e) => e.message,
            'message',
            'El nombre es obligatorio.',
          ),
        ),
      );
    });
  });
}
